### ISSUE TITLE
Unexpected ref safety errors when a discard is used as an `out` argument

### ISSUE THREAD

**Issue by AlekseyTs** [dotnet/roslyn#56587]
**Note after discussion this issue is meant to be tracking the discard case only**

```
using System;

partial class Test
{
    static void Main(){}
    
    void Test1()
    {
        Span<int> a = stackalloc int[10];
        Span<int> b;
        M(ref a, out b);
    }

    void M(ref Span<int> s, out Span<int> s1) => s1 = s;
}
```

Observed:
```
error CS8352: Cannot use local 'a' in this context because it may expose referenced variables outside of their declaration scope
error CS8350: This combination of arguments to 'Test.M(ref Span<int>, out Span<int>)' is disallowed because it may expose variables referenced by parameter 's' outside of their declaration scope
```

Expected: No errors.

The same behavior is around `out` discard:
```
        Span<int> a = stackalloc int[10];
        M(ref a, out _);
``` 

Note that, when local is initialized with stackalloc the errors go away:
```
        Span<int> a = stackalloc int[10];
        Span<int> b = stackalloc int[10];
        M(ref a, out b);
``` 
but there are errors in this case:
```
        Span<int> a = stackalloc int[10];
        Span<int> c = new int[10];
        M(ref a, out c);
```

**Comment by jaredpar** [comment]
This behavior is "By Design". The problem is generally referred to as the "method arguments must match" problem

https://github.com/dotnet/csharplang/blob/main/proposals/csharp-7.2/span-safety.md#method-arguments-must-match

The reason why they must match is because if they did not then developers could write code like the following: 

```c#
void M1(ref Span<int> s1)
{
    Span<int> s2 = stackalloc int[1];
    Swap(ref s1, ref s2);
}

void Swap(ref Span<int> x, ref Span<int> y)
{
    // This will effectively assign the stackalloc to the s1 parameter and allow it
    // to escape to the caller of M1
    ref x = ref y; 
}
```

This can be counter intuitive at a glance. Particularly because for the purposes of lifetime safety there is no difference between `out` and `ref`. Once you rewrite the sample to use `ref` in both cases it becomes a bit clearer why there is a problem.

**Comment by jaredpar** [comment]
Chatted offline a bit. The error is correct when an uninitialized local is used here because it allows for references to stack memory to escape. Example: 

```c#
    Span<int> Test1()
    {
        Span<int> a = stackalloc int[10];
        Span<int> b;
        M(ref a, out b);
        return b; // b points to the stack now
    }

    void M(ref Span<int> s, out Span<int> s1)
    {
        s1 = s; 
    }
```

In the case we use a discard though there is no such safety issue in theory. The value is never referenced hence can't be returned by the customer and hence it shouldn't contribute to the lifetime rules here:

```c#
    Span<int> Test2()
    {
        Span<int> a = stackalloc int[10];

        // The discard isn't accessible here, it can't be returned, it's uninitialized hence it should not contribute. 
        M(ref a, out _);
        ...
    }
```

I think that it would be good to change the language to allow this. 

In practice though the issue is a bit more complex. One of the issues we have to worry about with span safety is that `out` isn't enforceable, it's a C# construct only. Hence whenever we think about the semantics of out for span safety we have to think of it as `ref`. Hence it's always possible a swap method is implemented essentially like the following (in a language which supports `Span<T>` but does not respect `out`) 

```
  public void SillySwap([Out] ref Span<byte> s1, [Out] ref Span<byte> s2) { 
    Span<byte> temp = s1;
    s1 = s2;
    s2 = temp;
  }
```

Now this is a problem because we re-use discard variables as an implementation detail. Hence with this swap method we can return stack memory using discards. 

```c#
Span<byte> M() {
  Span<byte> a = stackalloc byte[42];
  SillySwap(out a, out _); // discard now references stack memory 
  Span<byte> b; 
  SillySwap(out _, out b); // b now references stack memory  
  return b;
} 
```

I don't think this is a blocker though, just more an item that we need to be thoughtful of. When enabling this we need to be careful to either zero out discards before passing them to `out` (for this scenario at least) or not re-using previous discards here.
